packages/toolboxes/src/cosmos/toolbox/thorchain.ts

```diff
@@ -165,26 +166,41 @@ async function signWithPrivateKey({
   return base64.encode(Buffer.concat([signature.r(32), signature.s(32)]));
 }
 
-export function createThorchainToolbox({
+export async function createThorchainToolbox({
   chain,
-  derivationPath: paramsDerivationPath,
-  index = 0,
-  signer,
-}: Omit<CosmosToolboxParams, "chain"> & { chain: Chain.THORChain | Chain.Maya }) {
+  ...toolboxParams
+}: CosmosToolboxParams<Chain.THORChain | Chain.Maya>) {
   const nodeUrl = SKConfig.get("nodeUrls")[chain];
   const rpcUrl = SKConfig.get("rpcUrls")[chain];
   const { isStagenet } = SKConfig.get("envs");
-
   const isThorchain = chain === Chain.THORChain;
   const chainPrefix = `${isStagenet ? "s" : ""}${CosmosChainPrefixes[chain]}`;
-  const derivationPath = paramsDerivationPath || DerivationPath[chain];
 
-  const cosmosToolbox = createCosmosToolbox({
+  const index = "index" in toolboxParams ? toolboxParams.index || 0 : 0;
+
+  const derivationPath = derivationPathToString(
+    "derivationPath" in toolboxParams && toolboxParams.derivationPath
+      ? toolboxParams.derivationPath
+      : updateDerivationPath(NetworkDerivationPath[chain], { index }),
+  );
+
+  const cosmosToolbox = await createCosmosToolbox({
     chain,
-    derivationPath,
-    signer,
-    index,
+    ...toolboxParams,
   });
+
+  const signer =
+    "phrase" in toolboxParams && toolboxParams.phrase
+      ? await cosmosToolbox.getSignerFromPhrase({
+          phrase: toolboxParams.phrase,
+          derivationPath:
+            toolboxParams.derivationPath ||
+            updateDerivationPath(NetworkDerivationPath[chain], { index }),
+        })
+      : "signer" in toolboxParams
+        ? toolboxParams.signer
+        : undefined;
+
```

Comment:
That kinda grew. WDYT about moving that to separate function with guard if clauses? Could be little bit more readable (as we have there few branches & ternaries nested)